### ISSUE TITLE
non_blocking=True still fails, lacking corresponding test case

### ISSUE THREAD

**Issue by andrewilyas** [libffcv/ffcv#37]
The ``non_blocking=True`` option still leads to nonsense values being returned by the dataloader. I observed this issue as part of a much larger system but have been unable to isolate a test case, so putting here in case anyone has seen something similar, otherwise I will get back to looking for a test case in a few days.

The loader code (as part of the larger system) is:

```python
    return  Loader(data_path,
                batch_size=2048,
                num_workers=10,
                order=OrderOption.QUASI_RANDOM,
                indices=subset,
                drop_last=True,
                os_cache=True,
                pipelines={
                    'a': [NDArrayDecoder(), ToTensor(), ToDevice(ch.device('cuda:0'), non_blocking=False)],
                    'b': [NDArrayDecoder(), ToTensor(), ToDevice(ch.device('cuda:0'), non_blocking=False)],
                    'idx': [IntDecoder(), ToTensor(), Squeeze(), ToDevice(ch.device('cuda:0'), non_blocking=False)]
                })
```

And the corresponding loop that fails is simply:

```python
a_table = ch.zeros(a_shape, dtype=ch.float32)
    w_grad_avg = ch.zeros_like(weight)
    b_grad_avg = ch.zeros_like(bias)

    done_inds = ch.zeros(num_outputs, dtype=ch.bool)
    residual = ch.zeros((batch_size, num_outputs), dtype=ch.float32).to(weight.device)

    w_saga = ch.zeros_like(weight)

    mm_mu, mm_sig = pct, np.sqrt(pct * (1-pct))
    for t in range(nepochs):
        total_loss = 0
        iterator = tqdm(loader)
        for bool_X, y, idx in iterator:
            # BS x NUM_CLASSES
            a_prev = a_table[idx].cuda(non_blocking=True)

            X = bool_X.float()
            bool_X.logical_not_()

            X -= mm_mu
            X /= mm_sig
            B = bool_X.sum(0) if train_mode else batch_size

            ch.mm(X, weight, out=residual)
            residual += bias
            residual -= y

            if train_mode: residual *= bool_X

            residual -= a_prev
            ch.mm(X.T, residual, out=w_saga)
            w_saga /= B
            w_saga += w_grad_avg
```

Replacing the loop with a ``pass`` fixes the issue, but none of the operations in the loop seem particularly important. Also changing non_blocking to false fixes it, so seems likely the problem is there.

**Comment by lykius** [comment]
Hi @andrewilyas,
`ffcv` is really promising, looking forward to its future!

I'm testing it in a simple application and it looks to me that there's still some problem with the `non_blocking` feature:

- If I use `ToDevice(0, non_blocking=True)`, I get some CUDA assert complaining that my labels are outside the range [0, n_classes].
- If I use `ToDevice(0, non_blocking=False)` everything works fine.
- If I don't use `ToDevice()` at all but use the standard `.cuda()` from pytorch, everything works fine.

Do you have any idea of what's going on?
I'm using version 0.0.2.

**Comment by GuillaumeLeclerc** [comment]
Sorry you are experiencing issues  Can you try v 0.0.3rc1 ?

Do you have an example that we could use to reproduce the bug ?

**Comment by lykius** [comment]
Thanks for the reply @GuillaumeLeclerc,

I'll try version 0.0.3rc1 tomorrow in the morning and I'll see if I can prepare a minimum working example.

**Comment by lykius** [comment]
It looks like the issue is not present in version 0.0.3rc1.

**Comment by andrewilyas** [comment]
Great to hear!

**Comment by GuillaumeLeclerc** [comment]
@lykius This is now in the 0.0.3 final release no need for the RC one.

**Comment by lykius** [comment]
Thanks guys!
